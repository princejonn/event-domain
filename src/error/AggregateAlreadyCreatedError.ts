import { DomainError } from "./DomainError";

export class AggregateAlreadyCreatedError extends DomainError {
  constructor() {
    super("Aggregate has already been created", { permanent: true });
  }
}
