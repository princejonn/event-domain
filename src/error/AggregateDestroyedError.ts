import { DomainError } from "./DomainError";

export class AggregateDestroyedError extends DomainError {
  constructor() {
    super("Aggregate is destroyed", { permanent: true });
  }
}
