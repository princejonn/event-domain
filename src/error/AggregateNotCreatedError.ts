import { DomainError } from "./DomainError";

export class AggregateNotCreatedError extends DomainError {
  constructor() {
    super("Aggregate has not been created", { permanent: true });
  }
}
