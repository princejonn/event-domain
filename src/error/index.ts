export * from "./AggregateAlreadyCreatedError";
export * from "./AggregateDestroyedError";
export * from "./AggregateNotCreatedError";
export * from "./AggregateNotDestroyedError";
export * from "./CausationMissingEventsError";
export * from "./CommandSchemaValidationError";
export * from "./ConcurrencyError";
export * from "./DomainError";
export * from "./EventStreamRevisionError";
export * from "./HandlerNotRegisteredError";
export * from "./IllegalEntityChangeError";
export * from "./MessageTypeError";
export * from "./MongoDuplicateKeyError";
export * from "./MongoNotUpdatedError";
export * from "./SagaAlreadyCreatedError";
export * from "./SagaCausationError";
export * from "./SagaDestroyedError";
export * from "./SagaNotCreatedError";
export * from "./SagaNotInStoreError";
export * from "./SagaRevisionError";
export * from "./ViewAlreadyCreatedError";
export * from "./ViewCausationError";
export * from "./ViewDestroyedError";
export * from "./ViewNotCreatedError";
export * from "./ViewRevisionError";
