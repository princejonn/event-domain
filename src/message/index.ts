export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
