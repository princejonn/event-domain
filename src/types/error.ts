import { LindormErrorOptions } from "@lindorm-io/errors";

export interface DomainErrorOptions extends LindormErrorOptions {
  permanent: boolean;
}
