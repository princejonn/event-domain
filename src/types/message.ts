import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
