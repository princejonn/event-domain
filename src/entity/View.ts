import { DomainEvent } from "../message";
import { IllegalEntityChangeError, ViewDestroyedError } from "../error";
import { Logger } from "@lindorm-io/winston";
import { MetaState } from "../enum";
import { assertCamelCase, validateSchemaSync } from "../util";
import { cloneDeep, find, get, isEqual, isMatch, remove, set, some } from "lodash";
import { isAfter, parseJSON } from "date-fns";
import {
  SagaIdentifier,
  IView,
  ViewOptions,
  AnyType,
  GenericRecord,
  ViewData,
} from "../types";
import {
  VIEW_ADD_FIELD_SCHEMA,
  VIEW_REMOVE_FIELD_WHERE_EQUAL_SCHEMA,
  VIEW_REMOVE_FIELD_WHERE_MATCH_SCHEMA,
  VIEW_SET_STATE_SCHEMA,
} from "../schema";

export class View<State> implements IView<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _causationList: Array<string>;
  private readonly _metaState: Record<string, any>;
  private readonly _revision: number;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;

  private readonly logger: Logger;

  constructor(options: ViewOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["View"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._causationList = options.causationList || [];
    this._destroyed = options.destroyed || false;
    this._metaState = options.metaState || {};
    this._revision = options.revision || 0;
    this._state = options.state || {};
  }

  public get causationList(): Array<string> {
    return this._causationList;
  }
  public set causationList(_) {
    throw new IllegalEntityChangeError();
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get metaState(): Record<string, any> {
    return this._metaState;
  }
  public set metaState(_) {
    throw new IllegalEntityChangeError();
  }

  public get revision(): number {
    return this._revision;
  }
  public set revision(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(v: GenericRecord<State>) {
    throw new IllegalEntityChangeError();
  }

  public addField(causation: DomainEvent, path: string, value: AnyType): void {
    this.logger.debug("adding field", { causation, path, value });

    validateSchemaSync(VIEW_ADD_FIELD_SCHEMA, { path, value, causation });

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const timestamp = causation.timestamp;
    const meta = get(this._metaState, path, []) as Array<any>;
    const field = get(this._state, path, []) as Array<any>;
    const exists = find(field, (item) => isEqual(item, value));

    const hasMoreRecentChange = some(
      meta,
      (item) =>
        isEqual(item[MetaState.VALUE], value) &&
        isAfter(parseJSON(item[MetaState.TIMESTAMP]), timestamp),
    );

    if (hasMoreRecentChange) {
      return;
    }

    remove(meta, (item) => isEqual(item[MetaState.VALUE], value));

    meta.push({
      [MetaState.REMOVED]: false,
      [MetaState.TIMESTAMP]: timestamp,
      [MetaState.VALUE]: value,
    });

    if (!exists) {
      field.push(value);
    }

    set(this._state, path, field);
    set(this._metaState, path, meta);
  }

  public destroy(): void {
    this.logger.debug("destroying");

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    this._destroyed = true;
  }

  public removeFieldWhereEqual(
    causation: DomainEvent,
    path: string,
    value: AnyType,
  ): void {
    this.logger.debug("removing field where equal", { causation, path, value });

    validateSchemaSync(VIEW_REMOVE_FIELD_WHERE_EQUAL_SCHEMA, { path, value, causation });

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const timestamp = causation.timestamp;
    const meta = get(this._metaState, path, []) as Array<any>;
    const field = get(this._state, path, []) as Array<any>;
    const exists = find(field, (item) => isEqual(item, value));

    if (!exists) {
      throw new Error("No existing value can be found");
    }

    const hasMoreRecentChange = some(
      meta,
      (item) =>
        isEqual(item[MetaState.VALUE], value) &&
        isAfter(parseJSON(item[MetaState.TIMESTAMP]), timestamp),
    );

    if (hasMoreRecentChange) {
      return;
    }

    remove(meta, (item) => isEqual(item[MetaState.VALUE], value));
    remove(field, (item) => isEqual(item, value));

    meta.push({
      [MetaState.REMOVED]: true,
      [MetaState.TIMESTAMP]: timestamp,
      [MetaState.VALUE]: exists,
    });

    set(this._state, path, field);
    set(this._metaState, path, meta);
  }

  public removeFieldWhereMatch(
    causation: DomainEvent,
    path: string,
    value: Record<string, any>,
  ): void {
    this.logger.debug("removing field where matching", { causation, path, value });

    validateSchemaSync(VIEW_REMOVE_FIELD_WHERE_MATCH_SCHEMA, { path, value, causation });

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const timestamp = causation.timestamp;
    const meta = get(this._metaState, path, []) as Array<any>;
    const field = get(this._state, path, []) as Array<any>;
    const exists = find<Record<string, any>>(field, value);

    if (!exists) {
      throw new Error("No matching value can be found");
    }

    const hasMoreRecentChange = some(
      meta,
      (item) =>
        isMatch(item[MetaState.VALUE], value) &&
        isAfter(parseJSON(item[MetaState.TIMESTAMP]), timestamp),
    );

    if (hasMoreRecentChange) {
      return;
    }

    remove(meta, (item) => isMatch(item[MetaState.VALUE], value));
    remove(field, (item) => isMatch(item, value));

    meta.push({
      [MetaState.REMOVED]: true,
      [MetaState.TIMESTAMP]: timestamp,
      [MetaState.VALUE]: exists,
    });

    set(this._state, path, field);
    set(this._metaState, path, meta);
  }

  public setState(causation: DomainEvent, path: string, value: AnyType): void {
    this.logger.debug("setting state", { causation, path, value });

    validateSchemaSync(VIEW_SET_STATE_SCHEMA, { path, value, causation });

    if (this._destroyed) {
      throw new ViewDestroyedError();
    }

    const timestamp = causation.timestamp;
    const meta = get(this.metaState, path, { timestamp }) as Record<string, any>;
    const hasMoreRecentChange = isAfter(parseJSON(meta[MetaState.TIMESTAMP]), timestamp);

    if (hasMoreRecentChange) {
      return;
    }

    set(this._state, path, value);
    set(this._metaState, path, {
      [MetaState.REMOVED]: false,
      [MetaState.TIMESTAMP]: timestamp,
      [MetaState.VALUE]: value,
    });
  }

  public toJSON(): ViewData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      causationList: cloneDeep(this.causationList),
      destroyed: this.destroyed,
      metaState: cloneDeep(this.metaState),
      revision: this.revision,
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: SagaIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
