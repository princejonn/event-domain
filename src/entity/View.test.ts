import MockDate from "mockdate";
import { Command, DomainEvent } from "../message";
import { IllegalEntityChangeError, ViewDestroyedError } from "../error";
import { View } from "./View";
import { aggregateIdentifier, logger, viewIdentifier } from "../test";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("View", () => {
  const event1 = new DomainEvent(
    {
      id: "15728f08-9693-4c9d-bb36-52614c189c96",
      name: "eventName",
      aggregate: aggregateIdentifier,
      data: { value: "hello world" },
      timestamp: new Date("2022-01-01T08:00:00.000Z"),
    },
    new Command({
      id: "d470745c-81c9-41b9-b6f0-e496ffbb77d2",
      name: "commandName",
      aggregate: aggregateIdentifier,
    }),
  );

  const event2 = new DomainEvent(
    {
      id: "1397365b-ffaa-46f5-b91b-6cfd06e807ed",
      name: "eventName",
      aggregate: aggregateIdentifier,
      data: { value: "foo bar" },
      timestamp: new Date("2022-01-01T08:01:00.000Z"),
    },
    new Command({
      id: "d797feea-a696-46fb-a05b-386a569038b8",
      name: "commandName",
      aggregate: aggregateIdentifier,
    }),
  );

  const event3 = new DomainEvent(
    {
      id: "6076bb19-cde8-4905-975a-8d2dbea5c620",
      name: "eventName",
      aggregate: aggregateIdentifier,
      data: { value: "foo bar" },
      timestamp: new Date("2022-01-01T08:02:00.000Z"),
    },
    new Command({
      id: "96b6139a-091b-46d4-8b4a-b4d985503002",
      name: "commandName",
      aggregate: aggregateIdentifier,
    }),
  );

  describe("when creating a view with non camel case name", () => {
    it("should throw on PascalCase", () => {
      expect(() => new View({ ...viewIdentifier, name: "TestView" }, logger)).toThrow(
        Error,
      );
    });

    it("should throw on snake_case", () => {
      expect(() => new View({ ...viewIdentifier, name: "test_view" }, logger)).toThrow(
        Error,
      );
    });

    it("should throw on kebab-case", () => {
      expect(() => new View({ ...viewIdentifier, name: "test-view" }, logger)).toThrow(
        Error,
      );
    });
  });

  describe("when setting state", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.setState(event1, "path1.path2", event1.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should set state", () => {
      expect(view.state.path1.path2).toStrictEqual("hello world");
    });

    it("should set meta state", () => {
      expect(view.metaState.path1.path2).toStrictEqual({
        r: false,
        t: new Date("2022-01-01T08:00:00.000Z"),
        v: "hello world",
      });
    });
  });

  describe("when overwriting state in order", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.setState(event1, "path1", event1.data.value);
      view.setState(event2, "path1", event2.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should set state to the most recent value", () => {
      expect(view.state.path1).toStrictEqual("foo bar");
    });

    it("should set meta state to the most recent event", () => {
      expect(view.metaState.path1).toStrictEqual({
        r: false,
        t: new Date("2022-01-01T08:01:00.000Z"),
        v: "foo bar",
      });
    });
  });

  describe("when overwriting state out of order", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.setState(event2, "path1", event2.data.value);
      view.setState(event1, "path1", event1.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should set state to the most recent value", () => {
      expect(view.state.path1).toStrictEqual("foo bar");
    });

    it("should set meta state to the most recent event", () => {
      expect(view.metaState.path1).toStrictEqual({
        r: false,
        t: new Date("2022-01-01T08:01:00.000Z"),
        v: "foo bar",
      });
    });
  });

  describe("when adding an item to an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", event1.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array", () => {
      expect(view.state.path1).toStrictEqual(["hello world"]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: false,
          t: new Date("2022-01-01T08:00:00.000Z"),
          v: "hello world",
        },
      ]);
    });
  });

  describe("when adding the same item to an array twice", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", event1.data.value);
      view.addField(event2, "path1", event1.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array once", () => {
      expect(view.state.path1).toStrictEqual(["hello world"]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: false,
          t: new Date("2022-01-01T08:01:00.000Z"),
          v: "hello world",
        },
      ]);
    });
  });

  describe("when adding the same object to an array twice", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", { hello: "world" });
      view.addField(event2, "path1", { hello: "world" });
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array once", () => {
      expect(view.state.path1).toStrictEqual([{ hello: "world" }]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: false,
          t: new Date("2022-01-01T08:01:00.000Z"),
          v: { hello: "world" },
        },
      ]);
    });
  });

  describe("when removing an existing item from an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", event1.data.value);
      view.removeFieldWhereEqual(event2, "path1", event1.data.value);
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array once", () => {
      expect(view.state.path1).toStrictEqual([]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: true,
          t: new Date("2022-01-01T08:01:00.000Z"),
          v: "hello world",
        },
      ]);
    });
  });

  describe("when removing a non-existing item from an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() =>
        view.removeFieldWhereEqual(event1, "path1", event1.data.value),
      ).toThrow();
    });
  });

  describe("when removing a matching item from an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", { id: 1, hello: "world" });
      view.removeFieldWhereMatch(event2, "path1", { id: 1 });
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array once", () => {
      expect(view.state.path1).toStrictEqual([]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: true,
          t: new Date("2022-01-01T08:01:00.000Z"),
          v: { id: 1, hello: "world" },
        },
      ]);
    });
  });

  describe("when removing a non-matching item from an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => view.removeFieldWhereMatch(event1, "path1", { id: 1 })).toThrow();
    });
  });

  describe("when adding a removed item to an array", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", "hello world");
      view.removeFieldWhereEqual(event2, "path1", "hello world");
      view.addField(event3, "path1", "hello world");
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array", () => {
      expect(view.state.path1).toStrictEqual(["hello world"]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: false,
          t: new Date("2022-01-01T08:02:00.000Z"),
          v: "hello world",
        },
      ]);
    });
  });

  describe("when adding a removed item to an array out of order", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.addField(event1, "path1", "hello world");
      view.addField(event3, "path1", "hello world");
      view.removeFieldWhereEqual(event2, "path1", "hello world");
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should add the value to the array", () => {
      expect(view.state.path1).toStrictEqual(["hello world"]);
    });

    it("should set meta state", () => {
      expect(view.metaState.path1).toStrictEqual([
        {
          r: false,
          t: new Date("2022-01-01T08:02:00.000Z"),
          v: "hello world",
        },
      ]);
    });
  });

  describe("when destroying a view", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.destroy();
    });

    it("should match snapshot", () => {
      expect(view).toMatchSnapshot();
    });

    it("should set the view as destroyed", () => {
      expect(view.destroyed).toBe(true);
    });
  });

  describe("when destroying a destroyed view", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);

      view.destroy();
    });

    it("should throw", () => {
      expect(() => view.destroy()).toThrow(ViewDestroyedError);
    });
  });

  describe("when trying to set causationList", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        view.causationList = [];
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set destroyed", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        view.destroyed = true;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set metaState", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        view.metaState = {};
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set revision", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        view.revision = 2;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set state", () => {
    let view: View<any>;

    beforeAll(() => {
      view = new View(viewIdentifier, logger);
    });

    it("should throw", () => {
      expect(() => {
        view.state = {};
      }).toThrow(IllegalEntityChangeError);
    });
  });
});
