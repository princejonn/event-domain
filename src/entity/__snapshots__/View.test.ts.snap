// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View when adding a removed item to an array out of order should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": false,
        "t": 2022-01-01T08:02:00.000Z,
        "v": "hello world",
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [
      "hello world",
    ],
  },
}
`;

exports[`View when adding a removed item to an array should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": false,
        "t": 2022-01-01T08:02:00.000Z,
        "v": "hello world",
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [
      "hello world",
    ],
  },
}
`;

exports[`View when adding an item to an array should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": false,
        "t": 2022-01-01T08:00:00.000Z,
        "v": "hello world",
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [
      "hello world",
    ],
  },
}
`;

exports[`View when adding the same item to an array twice should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": false,
        "t": 2022-01-01T08:01:00.000Z,
        "v": "hello world",
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [
      "hello world",
    ],
  },
}
`;

exports[`View when adding the same object to an array twice should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": false,
        "t": 2022-01-01T08:01:00.000Z,
        "v": Object {
          "hello": "world",
        },
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [
      Object {
        "hello": "world",
      },
    ],
  },
}
`;

exports[`View when destroying a view should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": true,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {},
  "name": "viewName",
  "revision": 0,
  "state": Object {},
}
`;

exports[`View when overwriting state in order should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Object {
      "r": false,
      "t": 2022-01-01T08:01:00.000Z,
      "v": "foo bar",
    },
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": "foo bar",
  },
}
`;

exports[`View when overwriting state out of order should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Object {
      "r": false,
      "t": 2022-01-01T08:01:00.000Z,
      "v": "foo bar",
    },
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": "foo bar",
  },
}
`;

exports[`View when removing a matching item from an array should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": true,
        "t": 2022-01-01T08:01:00.000Z,
        "v": Object {
          "hello": "world",
          "id": 1,
        },
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [],
  },
}
`;

exports[`View when removing an existing item from an array should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Array [
      Object {
        "r": true,
        "t": 2022-01-01T08:01:00.000Z,
        "v": "hello world",
      },
    ],
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Array [],
  },
}
`;

exports[`View when setting state should match snapshot 1`] = `
Object {
  "causationList": Array [],
  "context": "viewContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "metaState": Object {
    "path1": Object {
      "path2": Object {
        "r": false,
        "t": 2022-01-01T08:00:00.000Z,
        "v": "hello world",
      },
    },
  },
  "name": "viewName",
  "revision": 0,
  "state": Object {
    "path1": Object {
      "path2": "hello world",
    },
  },
}
`;
